In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
1,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
2,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
3,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
4,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# BORUTA

## 1. Baseline

In [10]:
!pip install boruta

  Using cached Boruta-0.4.3-py3-none-any.whl.metadata (8.8 kB)
Using cached Boruta-0.4.3-py3-none-any.whl (57 kB)


In [11]:
# =============================================================================
# BORUTA BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from boruta import BorutaPy
from sklearn.ensemble import RandomForestClassifier
import pandas as pd
import numpy as np


# =============================================================================
# BORUTA FEATURE SELECTION
# =============================================================================

# Random Forest used internally by Boruta
rf_boruta = RandomForestClassifier(
    n_estimators=200,
    class_weight="balanced",
    random_state=42,
    n_jobs=-1
)


# Initialize Boruta
boruta_selector = BorutaPy(
    estimator=rf_boruta,

    n_estimators="auto",

    verbose=0,

    random_state=42,

    max_iter=100,

    perc=100
)


# -------------------------------------------------------------------------
# Fit Boruta ONLY on training data
# -------------------------------------------------------------------------

boruta_selector.fit(
    X_train_scaled.values,
    y_train.values
)


# =============================================================================
# CREATE BORUTA RANKING
# =============================================================================

boruta_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "Boruta Rank": boruta_selector.ranking_,

    "Confirmed": boruta_selector.support_,

    "Tentative": boruta_selector.support_weak_

})


# -------------------------------------------------------------------------
# Sort by Boruta rank
# -------------------------------------------------------------------------

boruta_scores = boruta_scores.sort_values(
    by=["Boruta Rank", "Feature"]
).reset_index(drop=True)


print("=" * 70)
print("BORUTA FEATURE RANKING")
print("=" * 70)

print(boruta_scores)


# =============================================================================
# TOP-K FEATURE SELECTION
# =============================================================================

k_values = [5, 10, 15, 20]

boruta_results = {}


for k in k_values:

    # ---------------------------------------------------------------------
    # Select top-k features according to Boruta ranking
    # ---------------------------------------------------------------------

    top_features = (
        boruta_scores
        .head(k)["Feature"]
        .tolist()
    )


    # ---------------------------------------------------------------------
    # Boolean mask for selecting columns
    # ---------------------------------------------------------------------

    mask = X_train_scaled.columns.isin(
        top_features
    )


    # ---------------------------------------------------------------------
    # Select training and testing features
    # ---------------------------------------------------------------------

    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    # ---------------------------------------------------------------------
    # Store results
    # ---------------------------------------------------------------------

    boruta_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": boruta_scores.head(k)

    }


    # ---------------------------------------------------------------------
    # Display
    # ---------------------------------------------------------------------

    print("=" * 70)
    print(f"TOP {k} BORUTA FEATURES")
    print("=" * 70)

    print(top_features)

    print("\nBoruta Ranking:")
    print(
        boruta_scores.head(k)
    )

BORUTA FEATURE RANKING
                    Feature  Boruta Rank  Confirmed  Tentative
0                area error            1       True      False
1           concavity error            1       True      False
2                 mean area            1       True      False
3          mean compactness            1       True      False
4       mean concave points            1       True      False
5            mean concavity            1       True      False
6            mean perimeter            1       True      False
7               mean radius            1       True      False
8           mean smoothness            1       True      False
9              mean texture            1       True      False
10          perimeter error            1       True      False
11             radius error            1       True      False
12               worst area            1       True      False
13        worst compactness            1       True      False
14     worst concave points     

In [12]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {
    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,

        eval_metric="logloss",
        scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum()
    )
}

In [13]:
# =============================================================================
# MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


results = []


for k in k_values:

    Xtr = boruta_results[k]["X_train"]

    Xts = boruta_results[k]["X_test"]

    features = ", ".join(
        boruta_results[k]["features"]
    )


    # ---------------------------------------------------------------------
    # Evaluate every model
    # ---------------------------------------------------------------------

    for model_name, model in models.items():

        # Fit model
        model.fit(
            Xtr,
            y_train
        )


        # Predictions
        y_pred = model.predict(
            Xts
        )


        # Probability predictions
        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # Store metrics
        results.append({

            "Method": "Boruta",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


# =============================================================================
# FINAL BORUTA RESULTS TABLE
# =============================================================================

boruta_results_df = pd.DataFrame(
    results
)


display(
    boruta_results_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Boruta,5,Logistic Regression,"area error, concavity error, mean area, mean c...",0.912281,0.984375,0.875000,0.926471,0.980159
1,Boruta,5,Random Forest,"area error, concavity error, mean area, mean c...",0.929825,0.970588,0.916667,0.942857,0.986442
2,Boruta,5,XGBoost,"area error, concavity error, mean area, mean c...",0.921053,0.943662,0.930556,0.937063,0.981151
3,Boruta,10,Logistic Regression,"area error, concavity error, mean area, mean c...",0.894737,0.968750,0.861111,0.911765,0.987765
4,Boruta,10,Random Forest,"area error, concavity error, mean area, mean c...",0.956140,0.971831,0.958333,0.965035,0.989914
5,Boruta,10,XGBoost,"area error, concavity error, mean area, mean c...",0.938596,0.957746,0.944444,0.951049,0.990410
6,Boruta,15,Logistic Regression,"area error, concavity error, mean area, mean c...",0.938596,0.971014,0.930556,0.950355,0.994378
7,Boruta,15,Random Forest,"area error, concavity error, mean area, mean c...",0.956140,0.971831,0.958333,0.965035,0.994213
8,Boruta,15,XGBoost,"area error, concavity error, mean area, mean c...",0.964912,0.959459,0.986111,0.972603,0.994048
9,Boruta,20,Logistic Regression,"area error, concavity error, mean area, mean c...",0.956140,0.985507,0.944444,0.964539,0.997024


In [14]:
boruta_results_df.to_csv(
    "../../../results/breast_cancer/baseline_results/boruta_baseline_results.csv",
    index=False
)

## 2. DODA

In [15]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [16]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [17]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [18]:
# =============================================================================
# BORUTA + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from boruta import BorutaPy


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN BORUTA + DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"BORUTA + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # Random Forest estimator used by Boruta
    # ---------------------------------------------------------------------

    rf_boruta = RandomForestClassifier(
        n_estimators=200,
        class_weight="balanced",
        random_state=42,
        n_jobs=-1
    )


    # ---------------------------------------------------------------------
    # Boruta operator
    # ---------------------------------------------------------------------

    boruta = BorutaPy(
        estimator=rf_boruta,
        n_estimators="auto",
        max_iter=100,
        perc=100,
        random_state=42,
        verbose=0
    )


    # ---------------------------------------------------------------------
    # Wrap Boruta as a DODA operator
    # ---------------------------------------------------------------------

    boruta_operator = SklearnAdapter(
        boruta
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(
        operators=[boruta_operator],
        provider=provider,
        fusion=fusion,
        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = selector.get_selected_features()


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores": selector.raw_math_scores_,

        "math_scores": selector.math_scores_,

        "clinical_weights": selector.clinical_weights_,

        "final_scores": selector.final_scores_
    }


print("\n" + "=" * 70)
print("BORUTA + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


BORUTA + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000225C66D52B0>
DODA Pipeline Started
Running: BorutaPy
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of rankings: 30
Operator Scores:
{'BorutaPy_1': {'mean radius': np.float64(1.0), 'mean texture': np.float64(1.0), 'mean perimeter': np.float64(1.0), 'mean area': np.float64(1.0), 'mean smoothness': np.float64(1.0), 'mean compactness': np.float64(1.0), 'mean concavity': np.float64(1.0), 'mean concave points': np.float64(1.0), 'mean symmetry': np.float64(0.2), 'mean fractal dimension': np.float64(0.1111111111111111), 'radius error': np.float64(1.0), 'texture error': np.float64(0.2), 'perimeter error': np.float64(1.0), 'area error': np.float64(1.0), 'smoothness error': np

In [19]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
0,mean radius,1.000000,1.000000
1,mean texture,1.000000,1.000000
2,mean perimeter,1.000000,1.000000
3,mean area,1.000000,1.000000
4,mean smoothness,1.000000,1.000000
5,mean compactness,1.000000,1.000000
6,mean concavity,1.000000,1.000000
7,mean concave points,1.000000,1.000000
12,perimeter error,1.000000,1.000000
10,radius error,1.000000,1.000000


In [20]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
27,worst concave points,1.0
26,worst concavity,1.0
22,worst perimeter,1.0
23,worst area,1.0
20,worst radius,1.0
0,mean radius,0.9
3,mean area,0.9
6,mean concavity,0.9
7,mean concave points,0.9
2,mean perimeter,0.9


In [21]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
27,worst concave points,1.000000
26,worst concavity,1.000000
22,worst perimeter,1.000000
23,worst area,1.000000
20,worst radius,1.000000
0,mean radius,0.900000
3,mean area,0.900000
6,mean concavity,0.900000
7,mean concave points,0.900000
2,mean perimeter,0.900000


In [22]:
selector.get_selected_features()

['worst radius',
 'worst perimeter',
 'worst area',
 'worst concavity',
 'worst concave points',
 'mean radius',
 'mean perimeter',
 'mean area',
 'mean concavity',
 'mean concave points']

In [23]:
comparison = selector.compare_scores()

comparison

,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,worst radius,13,1.0000,1.0,1,1.0000,12
1,worst perimeter,15,1.0000,1.0,2,1.0000,13
2,worst area,16,1.0000,1.0,3,1.0000,13
3,worst concavity,19,1.0000,1.0,4,1.0000,15
4,worst concave points,20,1.0000,1.0,5,1.0000,15
5,mean radius,1,1.0000,0.9,6,0.9000,-5
6,mean perimeter,3,1.0000,0.9,7,0.9000,-4
7,mean area,4,1.0000,0.9,8,0.9000,-4
8,mean concavity,7,1.0000,0.9,9,0.9000,-2
9,mean concave points,8,1.0000,0.9,10,0.9000,-2


In [24]:
# =============================================================================
# BORUTA + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


doda_model_results = []


for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    # ---------------------------------------------------------------------
    # Evaluate all models
    # ---------------------------------------------------------------------

    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        doda_model_results.append({

            "Method": "Boruta + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# =============================================================================
# FINAL RESULTS TABLE
# =============================================================================

boruta_doda_results_df = pd.DataFrame(
    doda_model_results
)


display(
    boruta_doda_results_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Boruta + DODA,5,Logistic Regression,"worst radius, worst perimeter, worst area, wor...",0.921053,0.970149,0.902778,0.935252,0.988757
1,Boruta + DODA,5,Random Forest,"worst radius, worst perimeter, worst area, wor...",0.912281,0.942857,0.916667,0.929577,0.985946
2,Boruta + DODA,5,XGBoost,"worst radius, worst perimeter, worst area, wor...",0.938596,0.945205,0.958333,0.951724,0.978175
3,Boruta + DODA,10,Logistic Regression,"worst radius, worst perimeter, worst area, wor...",0.912281,0.969697,0.888889,0.927536,0.991402
4,Boruta + DODA,10,Random Forest,"worst radius, worst perimeter, worst area, wor...",0.947368,0.958333,0.958333,0.958333,0.990906
5,Boruta + DODA,10,XGBoost,"worst radius, worst perimeter, worst area, wor...",0.973684,0.972603,0.986111,0.979310,0.992063
6,Boruta + DODA,15,Logistic Regression,"worst radius, worst perimeter, worst area, wor...",0.947368,0.971429,0.944444,0.957746,0.993717
7,Boruta + DODA,15,Random Forest,"worst radius, worst perimeter, worst area, wor...",0.947368,0.958333,0.958333,0.958333,0.991237
8,Boruta + DODA,15,XGBoost,"worst radius, worst perimeter, worst area, wor...",0.938596,0.957746,0.944444,0.951049,0.992394
9,Boruta + DODA,20,Logistic Regression,"worst radius, worst perimeter, worst area, wor...",0.956140,0.985507,0.944444,0.964539,0.994378


In [25]:
boruta_doda_results_df.to_csv(
    "../../../results/breast_cancer/doda_results/boruta_doda_results.csv",
    index=False
)

In [26]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        boruta_results_df,
        boruta_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Boruta,5,Logistic Regression,"area error, concavity error, mean area, mean c...",0.912281,0.984375,0.875000,0.926471,0.980159
1,Boruta,5,Random Forest,"area error, concavity error, mean area, mean c...",0.929825,0.970588,0.916667,0.942857,0.986442
2,Boruta,5,XGBoost,"area error, concavity error, mean area, mean c...",0.921053,0.943662,0.930556,0.937063,0.981151
3,Boruta,10,Logistic Regression,"area error, concavity error, mean area, mean c...",0.894737,0.968750,0.861111,0.911765,0.987765
4,Boruta,10,Random Forest,"area error, concavity error, mean area, mean c...",0.956140,0.971831,0.958333,0.965035,0.989914
5,Boruta,10,XGBoost,"area error, concavity error, mean area, mean c...",0.938596,0.957746,0.944444,0.951049,0.990410
6,Boruta,15,Logistic Regression,"area error, concavity error, mean area, mean c...",0.938596,0.971014,0.930556,0.950355,0.994378
7,Boruta,15,Random Forest,"area error, concavity error, mean area, mean c...",0.956140,0.971831,0.958333,0.965035,0.994213
8,Boruta,15,XGBoost,"area error, concavity error, mean area, mean c...",0.964912,0.959459,0.986111,0.972603,0.994048
9,Boruta,20,Logistic Regression,"area error, concavity error, mean area, mean c...",0.956140,0.985507,0.944444,0.964539,0.997024


In [27]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'radius error', 'perimeter error', 'area error', 'concavity error', 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 'worst smoothness', 'worst compactness', 'worst concavity', 'worst concave points', 'worst symmetry', 'worst fractal dimension', 'concave points error', 'compactness error', 'mean symmetry', 'texture error', 'fractal dimension error', 'smoothness error', 'symmetry error', 'mean fractal dimension']

DODA Ranking:
['worst radius', 'worst perimeter', 'worst area', 'worst concavity', 'worst concave points', 'mean radius', 'mean perimeter', 'mean area', 'mean concavity', 'mean concave points', 'worst compactness', 'mean compactness', 'worst texture', 'worst symmetry', 'mean texture', 'radius error', 'perimeter error', 'area error', 'worst smoothness', 'mean smoothness', 'concavity error', 'wors

# 3. Boruta + DODA: Experiment Interpretation

## 3.1 Overview

This experiment compares a **data-driven Boruta baseline** with **Boruta + DODA**, where configured clinical knowledge is used to refine feature prioritization. The comparison evaluates whether DODA changes the Boruta-derived ranking and whether those changes improve, reduce, or maintain predictive performance across Top-5, Top-10, Top-15, and Top-20 feature subsets using Logistic Regression, Random Forest, and XGBoost.

The results should therefore be interpreted at two levels: **changes in feature prioritization** and **changes in downstream predictive performance**

## 3.2 Boruta Baseline Results

The Boruta baseline produced strong predictive performance across all Top-K feature subsets. Performance was not strictly improved by adding more features, indicating that the optimal subset size depended on the downstream model.

| Model | Strongest Boruta Top-K | Accuracy | Recall | F1-score | ROC-AUC |
|---|---:|---:|---:|---:|---:|
| Logistic Regression | Top-20 | 0.9561 | 0.9444 | 0.9645 | 0.9970 |
| Random Forest | Top-10 / Top-15 | 0.9561 | 0.9583 | 0.9650 | 0.9942–0.9899 |
| XGBoost | Top-15 | 0.9649 | 0.9861 | 0.9726 | 0.9940 |

**Observed patterns:**
- Logistic Regression reached its highest Accuracy and ROC-AUC with Top-20.
- Random Forest performed best in Accuracy with Top-10 and Top-15, while additional features did not improve Accuracy further.
- XGBoost achieved the strongest baseline Accuracy, Recall, and F1-score at Top-15.
- Top-10 to Top-20 generally produced stronger results than Top-5, although the best subset size remained model-dependent.

## 3.3 Boruta + DODA Results

After applying DODA, several configurations improved relative to the Boruta baseline, while others were reduced or remained comparable.

| Strong DODA Configuration | Accuracy | Recall | F1-score | ROC-AUC |
|---|---:|---:|---:|---:|
| Top-10 XGBoost | 0.9737 | 0.9861 | 0.9793 | 0.9921 |
| Top-20 Logistic Regression | 0.9561 | 0.9444 | 0.9645 | 0.9944 |
| Top-20 XGBoost | 0.9474 | 0.9722 | 0.9589 | 0.9967 |

**Observed results:**
- The clearest Accuracy and F1 improvement occurred for **XGBoost at Top-10**.
- Logistic Regression improved in several smaller feature subsets but did not exceed its baseline Accuracy at Top-20.
- Random Forest showed mixed results, with some reductions in Accuracy and ROC-AUC.
- Several configurations remained numerically close to the Boruta baseline, particularly at larger Top-K values.

Therefore, DODA did not produce a uniform improvement across all models and feature-set sizes.

## 3.4 Effect of DODA on Feature Ranking

Boruta assigned Rank 1 to a large group of confirmed features, including several mean, error, and worst measurements. This indicates limited differentiation among many of the Boruta-confirmed features because they shared the same top rank.

DODA substantially reprioritized the feature order by promoting clinically weighted **worst** measurements to the top of the ranking.

| Feature | Boruta Position/Status | DODA Position |
|---|---|---:|
| worst radius | Boruta Rank 1 | 1 |
| worst perimeter | Boruta Rank 1 | 2 |
| worst area | Boruta Rank 1 | 3 |
| worst concavity | Boruta Rank 1 | 4 |
| worst concave points | Boruta Rank 1 | 5 |
| mean radius | Boruta Rank 1 | 6 |
| mean perimeter | Boruta Rank 1 | 7 |
| mean area | Boruta Rank 1 | 8 |
| mean concavity | Boruta Rank 1 | 9 |
| mean concave points | Boruta Rank 1 | 10 |

**Interpretation:**
- DODA promoted clinically weighted features such as worst radius, worst perimeter, worst area, worst concavity, and worst concave points.
- These features were already Boruta-confirmed, so their promotion should be interpreted as **clinical reprioritization**, not as evidence that DODA statistically proved them more important.
- The main effect of DODA was a substantial change in ranking order and Top-K membership.
- Boruta's large group of Rank-1 features provided limited internal differentiation, while DODA imposed an additional prioritization layer based on configured clinical knowledge.

## 3.5 Top-K Feature Selection Comparison

### Top-5

**Boruta Top-5:**
- area error
- concavity error
- mean area
- mean compactness
- mean concave points

**Boruta + DODA Top-5:**
- worst radius
- worst perimeter
- worst area
- worst concavity
- worst concave points

The Top-5 subsets are completely different in membership and clearly show the effect of DODA's clinical prioritization.

### Top-10

Boruta retained primarily mean and error measurements, while DODA prioritized five worst measurements followed by mean radius, mean perimeter, mean area, mean concavity, and mean concave points.

The Top-10 membership therefore changed substantially rather than simply changing the order of an identical subset.

### Top-15

DODA retained several features also represented among Boruta-confirmed variables, but the ordering shifted toward clinically prioritized worst measurements and selected mean measurements.

### Top-20

At Top-20, both approaches covered a broader feature space, but the ranking order and selected subset membership still differed.

### Overall Interpretation

The notebook output shows `Top-K Same: False` for Top-5, Top-10, Top-15, and Top-20. Therefore, DODA did not merely reorder an identical Top-K subset. Instead, it changed feature membership at every evaluated Top-K level.

Feature-set stability increased conceptually as more features were included, but the actual Top-K outputs remained different across all four subset sizes.

## 3.6 Predictive Performance Comparison

The table below shows the change from **Boruta → Boruta + DODA**.

| Top-K | Model | Accuracy Change | Recall Change | F1 Change | ROC-AUC Change | Overall Effect |
|---|---|---:|---:|---:|---:|---|
| 5 | Logistic Regression | +0.0088 | +0.0278 | +0.0088 | +0.0086 | Improved |
| 5 | Random Forest | -0.0175 | 0.0000 | -0.0133 | -0.0005 | Reduced |
| 5 | XGBoost | +0.0175 | +0.0278 | +0.0147 | -0.0030 | Improved overall |
| 10 | Logistic Regression | +0.0175 | +0.0278 | +0.0158 | +0.0036 | Improved |
| 10 | Random Forest | -0.0088 | 0.0000 | -0.0067 | +0.0010 | Comparable / slightly reduced |
| 10 | XGBoost | +0.0351 | +0.0417 | +0.0283 | +0.0017 | Improved |
| 15 | Logistic Regression | +0.0088 | +0.0139 | +0.0074 | -0.0007 | Comparable / slightly improved |
| 15 | Random Forest | -0.0088 | 0.0000 | -0.0067 | +0.0013 | Comparable / slightly reduced |
| 15 | XGBoost | -0.0263 | -0.0417 | -0.0216 | -0.0017 | Reduced |
| 20 | Logistic Regression | 0.0000 | 0.0000 | 0.0000 | -0.0026 | Maintained / comparable |
| 20 | Random Forest | +0.0088 | +0.0278 | +0.0080 | 0.0000 | Improved |
| 20 | XGBoost | -0.0088 | 0.0000 | -0.0066 | +0.0020 | Comparable / slightly reduced |

**Overall pattern:** DODA had a clear model-dependent and Top-K-dependent effect. The strongest improvement occurred for XGBoost at Top-10, while some Random Forest and XGBoost configurations were reduced. Several larger feature subsets remained highly comparable to the Boruta baseline.

## 3.7 Logistic Regression Results

For Logistic Regression, DODA generally improved performance at Top-5, Top-10, and Top-15.

- **Best Boruta configuration:** Top-20 with Accuracy = 0.9561, F1-score = 0.9645, and ROC-AUC = 0.9970.
- **Best DODA Accuracy:** Top-20 with Accuracy = 0.9561.
- **Largest DODA improvement:** Top-10, where Accuracy increased from 0.8947 to 0.9123 and F1-score increased from 0.9118 to 0.9275.
- At Top-20, Accuracy, Recall, and F1-score were identical between Boruta and DODA, although ROC-AUC was slightly lower with DODA.

Overall, DODA improved smaller and medium-sized feature subsets for Logistic Regression, while the largest subset preserved essentially the same predictive performance.

## 3.8 Random Forest Results

Random Forest showed a mixed response to DODA-adjusted feature subsets.

- **Best Boruta Accuracy:** Top-10 and Top-15, both at 0.9561.
- **Best DODA Accuracy:** Top-20 at 0.9474.
- Top-5, Top-10, and Top-15 showed lower Accuracy and F1-score after DODA.
- Top-20 improved Accuracy from 0.9386 to 0.9474 and Recall from 0.9306 to 0.9583, while ROC-AUC remained identical at 0.9939.

These results indicate that Random Forest did not consistently benefit from the DODA-adjusted ranking. Larger Top-K values produced the strongest DODA result, but still did not exceed the strongest Boruta Accuracy.

## 3.9 XGBoost Results

XGBoost showed the strongest improvement from DODA at Top-10.

- **Best Boruta configuration:** Top-15 with Accuracy = 0.9649, Recall = 0.9861, F1-score = 0.9726, and ROC-AUC = 0.9940.
- **Best DODA Accuracy, Recall, and F1-score:** Top-10 with Accuracy = 0.9737, Recall = 0.9861, and F1-score = 0.9793.
- Top-5 also improved in Accuracy, Recall, and F1-score, although ROC-AUC was slightly lower.
- Top-15 showed a substantial reduction after DODA.
- Top-20 showed slightly lower Accuracy and F1-score but achieved the highest DODA ROC-AUC of 0.9967.

Therefore, XGBoost benefited from DODA at smaller feature subsets, particularly Top-10, but the effect depended strongly on feature-set size.

## 3.10 Best Performing Configurations

The strongest observed configurations across the experiment are shown below.

| Feature Selection Method | Top-K | Model | Accuracy | Recall | F1-score | ROC-AUC |
|---|---:|---|---:|---:|---:|---:|
| Boruta + DODA | 10 | XGBoost | 0.9737 | 0.9861 | 0.9793 | 0.9921 |
| Boruta | 15 | XGBoost | 0.9649 | 0.9861 | 0.9726 | 0.9940 |
| Boruta | 20 | Logistic Regression | 0.9561 | 0.9444 | 0.9645 | 0.9970 |
| Boruta + DODA | 20 | XGBoost | 0.9474 | 0.9722 | 0.9589 | 0.9967 |

### Best Metric-Specific Configurations

1. **Best Accuracy:** Boruta + DODA, Top-10, XGBoost = **0.9737**
2. **Best Recall:** A tie at **0.9861**, achieved by:
   - Boruta, Top-15, XGBoost
   - Boruta + DODA, Top-10, XGBoost
3. **Best F1-score:** Boruta + DODA, Top-10, XGBoost = **0.9793**
4. **Best ROC-AUC:** Boruta, Top-20, Logistic Regression = **0.9970**

No single configuration was best across all evaluation metrics.

## 3.11 Importance of Accuracy, Recall, F1-score, and ROC-AUC

The four metrics provide different perspectives on model performance:

- **Accuracy** measures overall classification correctness.
- **Recall** measures the ability to correctly identify positive cases under the experiment's target encoding.
- **F1-score** balances Precision and Recall.
- **ROC-AUC** measures the model's ability to discriminate between classes across decision thresholds.

In this experiment, the configuration with the highest Accuracy and F1-score, Boruta + DODA with Top-10 XGBoost, was not the same configuration with the highest ROC-AUC, Boruta with Top-20 Logistic Regression.

This demonstrates why performance should not be judged using Accuracy alone. Different feature-selection and model combinations may optimize different aspects of predictive performance.

## 3.12 Effect of Clinical Knowledge

DODA contributed a clinically informed prioritization layer to the Boruta-derived feature ranking.

The observed results show that:
- Clinically high-weighted worst measurements moved to the top of the DODA ranking.
- Several features already supported by Boruta remained important, but their ordering and Top-K membership changed.
- The resulting feature subsets differed from the Boruta Top-K subsets at all evaluated values of K.
- Changes in feature prioritization did not automatically translate into improved predictive performance.
- The effect depended on both the downstream model and the number of selected features.

The strongest observed benefit occurred for **XGBoost at Top-10**, while other configurations remained comparable or showed reduced performance. Therefore, clinical prioritization and predictive optimization were related in this experiment but were not equivalent objectives.

## 3.13 Important Observations

- DODA had a **model-dependent effect**: XGBoost showed the strongest improvement, while Random Forest showed several reductions.
- The effect was also **Top-K-dependent**: the same model could improve at one subset size and decrease at another.
- DODA substantially changed the feature ranking and Top-K membership, but these ranking changes did not always produce equally large changes in predictive performance.
- Logistic Regression generally benefited from DODA at smaller and medium-sized subsets, while its Top-20 performance was essentially preserved.
- Random Forest did not consistently benefit from DODA and performed best with the larger DODA subset at Top-20.
- XGBoost achieved the strongest overall Accuracy and F1-score with DODA at Top-10.
- Some configurations showed meaningful changes in feature selection while predictive performance remained relatively comparable.
- The experiment demonstrates that the same feature-ranking modification can produce different responses across classifiers.

## 3.14 Main Findings

1. Boruta provided a strong data-driven baseline, with many confirmed features sharing Rank 1 and producing high predictive performance across all three classifiers.
2. DODA substantially refined the ranking by promoting clinically weighted worst measurements and changing Top-K feature membership.
3. Important data-driven features were not simply discarded; several Boruta-confirmed features remained prominent after DODA, although their priority changed.
4. Predictive performance was generally stable and remained high across both approaches, but the effect of DODA varied by classifier and Top-K size.
5. DODA did **not** consistently improve performance across all configurations.
6. The strongest observed benefit was Boruta + DODA with Top-10 features and XGBoost, which achieved the highest Accuracy and F1-score in the experiment.

## 3.15 Research Interpretation

> Does incorporating clinical knowledge into feature-selection rankings consistently improve or preserve predictive performance across datasets and selection strategies?

This Boruta experiment provides **one dataset-specific piece of evidence** for this broader question.

The results suggest that incorporating configured clinical knowledge can substantially alter feature prioritization while maintaining or, in some configurations, improving predictive performance. However, the effect was not consistent across all models or Top-K subsets.

In this experiment:
- The breast cancer dataset represents only one predictive setting.
- Boruta represents only one underlying feature-selection strategy.
- DODA's predictive effect depended on the downstream classifier.
- Feature-set size influenced whether performance improved, remained comparable, or decreased.
- Clinical prioritization and purely predictive optimization were related but not identical.

Therefore, this experiment supports further investigation rather than establishing a general rule that clinically informed feature ranking consistently improves predictive performance.

## 3.16 Important Limitations

- The experiment was conducted using a single breast cancer dataset.
- The findings are specific to the current train-test split, model settings, evaluation procedure, and Top-K values.
- No statistical significance testing was performed in the reported results.
- The experiment does not validate the approach across multiple diseases or independent datasets.
- Clinical weights are configured inputs, meaning their specification can directly influence the resulting DODA ranking.
- The results therefore cannot establish universal generalization or universal superiority of DODA.

## 3.17 Final Conclusion

In this breast cancer experiment, Boruta provided a strong data-driven baseline for feature selection, producing high predictive performance across Logistic Regression, Random Forest, and XGBoost.

DODA substantially refined the Boruta-derived prioritization by promoting clinically weighted features and changing Top-K feature membership at every evaluated subset size. However, these ranking changes did not produce a uniform predictive advantage.

The strongest observed improvement occurred with **Boruta + DODA using Top-10 features and XGBoost**, which achieved the highest Accuracy and F1-score. Other configurations remained comparable to the Boruta baseline, while some showed reduced performance.

Overall, the results suggest that combining data-driven feature selection with configured clinical knowledge can be useful as a feature-prioritization strategy, but its predictive effect depends on the classifier and the number of selected features. This experiment therefore provides dataset-specific evidence rather than demonstrating that DODA is universally superior, and broader validation across datasets and feature-selection strategies is required.